{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true,
    "execution": {
     "iopub.execute_input": "2022-01-07T18:18:54.728067Z",
     "iopub.status.busy": "2022-01-07T18:18:54.726515Z",
     "iopub.status.idle": "2022-01-07T18:20:12.220432Z",
     "shell.execute_reply": "2022-01-07T18:20:12.219128Z"
    },
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Attaching SeuratObject\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘glue’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘cowplot’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:patchwork’:\n",
      "\n",
      "    align_plots\n",
      "\n",
      "\n",
      "Loading required package: Matrix\n",
      "\n",
      "\n",
      "Attaching package: ‘Matrix’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:tidyr’:\n",
      "\n",
      "    expand, pack, unpack\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(Seurat)\n",
    "\n",
    "library(dplyr)\n",
    "library(glue)\n",
    "library(ggplot2)\n",
    "library(patchwork)\n",
    "library(cowplot)\n",
    "library(stringr)\n",
    "library(tidyr)\n",
    "library(\"dplyr\")\n",
    "library(velocyto.R)\n",
    "library(stringr)\n",
    "library(SeuratWrappers)\n",
    "library(DAseq)\n",
    "load(\"manual_label.RData\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-01-07T18:30:28.457002Z",
     "iopub.status.busy": "2022-01-07T18:30:28.421861Z",
     "iopub.status.idle": "2022-01-07T18:30:28.468786Z",
     "shell.execute_reply": "2022-01-07T18:30:28.467693Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "deg=c('Trim12a','Echdc2','Gm42418','Rpl35a','Rps29','Rps15a','Rpl28','Rpl32','Rpl37a','Rplp1','Rpl41','Rpl39','Rps12','Rps20','Trim34a','Rps26','Rpl34','Rps25','Fau','Rpl36','Rpl37','Rps28','Rpl38','Wdfy1','H2-Q7','Rpl22l1','Gstp1','Gm47283','Lmo2','Tspan32','H2-Eb1','Ndrg1','Cox6b1','Maf','H2-Q4','Arhgdib','Fads1','Apoe','Gdpd3','Tnfsf10','Jund','S100a11','Adipor2','E230032D23Rik','Smim5','Rbm3','Lgals1','Psmb8','Gnb4','Mettl7a1','Rsrp1','Tmed5','Ankrd12','H2-Q6','Sem1','Wnk1','Sh3bgrl3','Akap13','Tap1','Meg3','Clec2d','Gse1','Stat3','Snrpe','Cyp51','Atp5e','Mrpl54','Mpl','Cox7a2','Snrpf','Numa1','Ctsg','Angpt1','Atp1a1','Aplp2','Glg1','Rras','Tsc22d3','Insig1','Arid1b','Tagln2','Cox6c','Rb1cc1','Ncor1','Uqcrb','Stat1','Prtn3','Hlf','Ly6a','Scd2','Nop10','Rfk','Vgll4','Gclm','Eif4b','Prr13','Uqcc2','Mrpl23','Ssr4','Ndufv3','Ltb','Runx1','Kansl1l','Cox5b','Pdzk1ip1','Tuba1b','Ifi27','Sec61g','Ranbp1','Nrip1','Ly6e','Crip1','Mpo','Alad','Ost4','Muc13','Rps17','H2-K1','Tgm2','Rab6b','Plppr3','Idi1','Hmgcr','Rabac1','Gcnt2','Pdcd4','Apobec3','Usp8','Slc2a3','Osbpl8','Hint1','Etf1','0610009B22Rik','Chic2','Gpr171','Akirin2','Slc35c2','Trim56','Sptan1','Tsc22d1','Hipk1','Shisa5','Gon4l','Pabpn1','Sp140','Uvssa','Fdps','Pcnt','Sugt1','Dtx3l','Lars2','Tpst2','Iigp1','Ndufa13','H2-Ab1','Trim47','Cox8a','Atp5j2','Cbx3','Nfe2','Ldlr','Cd74','Nufip2','Nfkb1','Sypl','Camk2d','Madd','Cers4','Ddb1','Rap1b','Eps8','Atpif1','Foxp1','Ctla2a','Pgap2','Hnrnpd','H2-T22','Smarca2','Lbhd2','AW112010','Eif3d','2410015M20Rik','Ptprc','Snrpd2','2010107E04Rik','Cks2','Aes','Cbfa2t3','Cox7b','AI314180','Rgs1','Mcl1','H2-T23','Dars','Myb','Hbb-bt','Trim5','Zfp36','H2-DMa','Lamp2','Slfn5','Adgrl4','Jun','Setd1b','Arhgap45','Rpl36al','Pou2f2','Mif','Ighm','Txnip','Herc6','Elane','Mtdh','Plek','Gm4951','Stxbp4','Atp5g1','Kdelr1','Ifi203','Junb','Igtp','Rasa4','Cd44','Lcp1','Ifi44','Bst2','Mndal','Rps27l','Satb1','Uqcr11','Mrps28','Rps21','Echdc1','Rpl7l1','Vldlr','Ubb','H2-D1','Zfp395','Tnpo1','Edem2','Dalrd3','Rnf149','Selenot','Usp48','Phb','Dnase2a','Zfp800','Lman2','Gpd2','Mef2d','Rrbp1','Smim14','F630028O10Rik','Crim1','Dock10','4932438A13Rik','Nceh1','Tars','Dock8','Slc39a10','Gm2a','Maea','Calr','Samd9l','Sympk','Lfng','Uqcr10','Mien1','Iqgap2','Tmem176b','Mylip','Rbck1','Mprip','Mycbp2','Laptm5','D430042O09Rik','Zmym4','Ndufa3','Smg1','Hnrnph3','Snrpg','Wdr26','Phf12','Pik3ip1','Ndufa4','Prkag2','Ppp2r2a','Macf1','Fam107b','Inpp4b','Ccdc47','Rpl23','Rpl5','Erh','Bcl2','Abl1','Eci1','Nkg7','Myl10','Ptpn18','Pbx4','Igfbp4','Rcl1','Hsp90aa1','Slc25a5','Nfix','Mrps30','Fam126b','H2afy','Gm42047','Spc24','Rfc2','Xrn2','Chchd10','Amd1','Exo1','Cycs','Abhd14a','Fam208b','Cul3','Psmb9','Fam136a','Rpa2','Mrpl32','Wdr76','Sec23ip','Hras','Ppp4c','Btaf1','Ebp','Uhrf1','Rabgap1l','Racgap1','Krtcap2','Smarce1','Rpl27','Knl1','Msh2','Ahsa1','Wdr5','Pds5a','Nisch','Hspa9','Tmx4','Uqcrh','Rrp1b','Eif3g','Mrto4','Hus1','Ssrp1','Hmgb2','Glrx5','Cct5','Glipr1','Cdk6','Snu13','Ube2s','Psmd11','Hspa5','Cirbp','Sdf2l1','Nop56','Dynll1','Lsm8','Sap30','Zbtb20','Map2k3','Fkbp4','Trim59','Itpa','Clspn','Gm26917','Cd63','Cenpa','Slc25a4','Ube2c','Tmem238','Gimap6','Chaf1a','Ung','Cd27','Gnl3','Cdca3','Exosc10','Tm2d1','Usmg5','Ybx3','Mrpl15','Anxa5','Cdv3','Myl6','S100a8','Uqcrq','Zcchc6','Atp5k','Tpm4','Kcnq1ot1','Ucp2','Cenpw','Ndufa2','Tubb5','Slc14a1','Pgam1','Top2a','Pf4','Plac8','Rps7','Ifngr1','Rpl13a','Txn1','Itga4','Ube2b','Gm17590','Raf1','Ldah','Stom','Anapc16','Gm11808','Ccdc86','Acss2','Car1','Ccdc88a','Gm35082','Jak2','Akirin1','Mllt10','Oat','Fam96a','Scin','Msn','Rsu1','Hsdl1','Mdn1','Zbtb1','Cdk19','Trib2','Tspo2','Ogt','Kctd14','Tmem230','Top2b','Zfp217','Acaa2','Dhx40','S100a10','Atp8a1','Arid4a','Ptpre','Samd14','Hist1h1b','Tspo','Ddx6','Smc4','Mosmo','Abcb4','Ehmt2','Vim','Atp5l','Smap2','Ldhb','Tbcel','Letm2','Tgoln1','Patz1','Runx3','Ndufb2','Dbi','Aqp1','Hells','Ttc3','Galnt6','Adgrg1','Gadd45a','Npc2','Krit1','Ppp4r2','Spn','Capza2','Ash1l','Kdm5a','Cox7c','mt-Co3','Mt1','Elob','Map1lc3b','Fam117b','mt-Atp6','Lsm5','Vamp5','Atox1','Ccdc50','mt-Nd2','Bcl11a','Mt2','Icam4','Birc5','Ssbp1','Psat1','Nipbl','Kmt2a','Sphk1','Malat1','Rps27','Pdia6','Manf','Ciapin1','Hspd1','Lcn2','Dnajc3','Eno1','Cacybp','S100a9','Cd34','Sec61b','Mrpl33','Pim1','Cd52','Clec12a','Slpi','Ms4a3','Prss57','Mrpl52','Ubl5','Hbb-bs','Ctse','Cela1','AY036118','Hist1h2ap','Tmsb4x','Hist1h2ae')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true,
    "execution": {
     "iopub.execute_input": "2022-01-07T18:31:00.008536Z",
     "iopub.status.busy": "2022-01-07T18:31:00.007496Z",
     "iopub.status.idle": "2022-01-07T18:31:04.967685Z",
     "shell.execute_reply": "2022-01-07T18:31:04.964022Z"
    },
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table width=\"100%\" summary=\"page for FindMarkers {Seurat}\"><tr><td>FindMarkers {Seurat}</td><td style=\"text-align: right;\">R Documentation</td></tr></table>\n",
       "\n",
       "<h2>Gene expression markers of identity classes</h2>\n",
       "\n",
       "<h3>Description</h3>\n",
       "\n",
       "<p>Finds markers (differentially expressed genes) for identity classes\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Usage</h3>\n",
       "\n",
       "<pre>\n",
       "FindMarkers(object, ...)\n",
       "\n",
       "## Default S3 method:\n",
       "FindMarkers(\n",
       "  object,\n",
       "  slot = \"data\",\n",
       "  counts = numeric(),\n",
       "  cells.1 = NULL,\n",
       "  cells.2 = NULL,\n",
       "  features = NULL,\n",
       "  logfc.threshold = 0.25,\n",
       "  test.use = \"wilcox\",\n",
       "  min.pct = 0.1,\n",
       "  min.diff.pct = -Inf,\n",
       "  verbose = TRUE,\n",
       "  only.pos = FALSE,\n",
       "  max.cells.per.ident = Inf,\n",
       "  random.seed = 1,\n",
       "  latent.vars = NULL,\n",
       "  min.cells.feature = 3,\n",
       "  min.cells.group = 3,\n",
       "  pseudocount.use = 1,\n",
       "  fc.results = NULL,\n",
       "  densify = FALSE,\n",
       "  ...\n",
       ")\n",
       "\n",
       "## S3 method for class 'Assay'\n",
       "FindMarkers(\n",
       "  object,\n",
       "  slot = \"data\",\n",
       "  cells.1 = NULL,\n",
       "  cells.2 = NULL,\n",
       "  features = NULL,\n",
       "  logfc.threshold = 0.25,\n",
       "  test.use = \"wilcox\",\n",
       "  min.pct = 0.1,\n",
       "  min.diff.pct = -Inf,\n",
       "  verbose = TRUE,\n",
       "  only.pos = FALSE,\n",
       "  max.cells.per.ident = Inf,\n",
       "  random.seed = 1,\n",
       "  latent.vars = NULL,\n",
       "  min.cells.feature = 3,\n",
       "  min.cells.group = 3,\n",
       "  pseudocount.use = 1,\n",
       "  mean.fxn = NULL,\n",
       "  fc.name = NULL,\n",
       "  base = 2,\n",
       "  densify = FALSE,\n",
       "  ...\n",
       ")\n",
       "\n",
       "## S3 method for class 'DimReduc'\n",
       "FindMarkers(\n",
       "  object,\n",
       "  cells.1 = NULL,\n",
       "  cells.2 = NULL,\n",
       "  features = NULL,\n",
       "  logfc.threshold = 0.25,\n",
       "  test.use = \"wilcox\",\n",
       "  min.pct = 0.1,\n",
       "  min.diff.pct = -Inf,\n",
       "  verbose = TRUE,\n",
       "  only.pos = FALSE,\n",
       "  max.cells.per.ident = Inf,\n",
       "  random.seed = 1,\n",
       "  latent.vars = NULL,\n",
       "  min.cells.feature = 3,\n",
       "  min.cells.group = 3,\n",
       "  pseudocount.use = 1,\n",
       "  mean.fxn = rowMeans,\n",
       "  fc.name = NULL,\n",
       "  densify = FALSE,\n",
       "  ...\n",
       ")\n",
       "\n",
       "## S3 method for class 'Seurat'\n",
       "FindMarkers(\n",
       "  object,\n",
       "  ident.1 = NULL,\n",
       "  ident.2 = NULL,\n",
       "  group.by = NULL,\n",
       "  subset.ident = NULL,\n",
       "  assay = NULL,\n",
       "  slot = \"data\",\n",
       "  reduction = NULL,\n",
       "  features = NULL,\n",
       "  logfc.threshold = 0.25,\n",
       "  test.use = \"wilcox\",\n",
       "  min.pct = 0.1,\n",
       "  min.diff.pct = -Inf,\n",
       "  verbose = TRUE,\n",
       "  only.pos = FALSE,\n",
       "  max.cells.per.ident = Inf,\n",
       "  random.seed = 1,\n",
       "  latent.vars = NULL,\n",
       "  min.cells.feature = 3,\n",
       "  min.cells.group = 3,\n",
       "  pseudocount.use = 1,\n",
       "  mean.fxn = NULL,\n",
       "  fc.name = NULL,\n",
       "  base = 2,\n",
       "  densify = FALSE,\n",
       "  ...\n",
       ")\n",
       "</pre>\n",
       "\n",
       "\n",
       "<h3>Arguments</h3>\n",
       "\n",
       "<table summary=\"R argblock\">\n",
       "<tr valign=\"top\"><td><code>object</code></td>\n",
       "<td>\n",
       "<p>An object</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>...</code></td>\n",
       "<td>\n",
       "<p>Arguments passed to other methods and to specific DE methods</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>slot</code></td>\n",
       "<td>\n",
       "<p>Slot to pull data from; note that if <code>test.use</code> is &quot;negbinom&quot;, &quot;poisson&quot;, or &quot;DESeq2&quot;,\n",
       "<code>slot</code> will be set to &quot;counts&quot;</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>counts</code></td>\n",
       "<td>\n",
       "<p>Count matrix if using scale.data for DE tests. This is used for\n",
       "computing pct.1 and pct.2 and for filtering features based on fraction\n",
       "expressing</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>cells.1</code></td>\n",
       "<td>\n",
       "<p>Vector of cell names belonging to group 1</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>cells.2</code></td>\n",
       "<td>\n",
       "<p>Vector of cell names belonging to group 2</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>features</code></td>\n",
       "<td>\n",
       "<p>Genes to test. Default is to use all genes</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>logfc.threshold</code></td>\n",
       "<td>\n",
       "<p>Limit testing to genes which show, on average, at least\n",
       "X-fold difference (log-scale) between the two groups of cells. Default is 0.25\n",
       "Increasing logfc.threshold speeds up the function, but can miss weaker signals.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>test.use</code></td>\n",
       "<td>\n",
       "<p>Denotes which test to use. Available options are:\n",
       "</p>\n",
       "\n",
       "<ul>\n",
       "<li><p>&quot;wilcox&quot; : Identifies differentially expressed genes between two\n",
       "groups of cells using a Wilcoxon Rank Sum test (default)\n",
       "</p>\n",
       "</li>\n",
       "<li><p>&quot;bimod&quot; : Likelihood-ratio test for single cell gene expression,\n",
       "(McDavid et al., Bioinformatics, 2013)\n",
       "</p>\n",
       "</li>\n",
       "<li><p>&quot;roc&quot; : Identifies 'markers' of gene expression using ROC analysis.\n",
       "For each gene, evaluates (using AUC) a classifier built on that gene alone,\n",
       "to classify between two groups of cells. An AUC value of 1 means that\n",
       "expression values for this gene alone can perfectly classify the two\n",
       "groupings (i.e. Each of the cells in cells.1 exhibit a higher level than\n",
       "each of the cells in cells.2). An AUC value of 0 also means there is perfect\n",
       "classification, but in the other direction. A value of 0.5 implies that\n",
       "the gene has no predictive power to classify the two groups. Returns a\n",
       "'predictive power' (abs(AUC-0.5) * 2) ranked matrix of putative differentially\n",
       "expressed genes.\n",
       "</p>\n",
       "</li>\n",
       "<li><p>&quot;t&quot; : Identify differentially expressed genes between two groups of\n",
       "cells using the Student's t-test.\n",
       "</p>\n",
       "</li>\n",
       "<li><p>&quot;negbinom&quot; : Identifies differentially expressed genes between two\n",
       "groups of cells using a negative binomial generalized linear model.\n",
       "Use only for UMI-based datasets\n",
       "</p>\n",
       "</li>\n",
       "<li><p>&quot;poisson&quot; : Identifies differentially expressed genes between two\n",
       "groups of cells using a poisson generalized linear model.\n",
       "Use only for UMI-based datasets\n",
       "</p>\n",
       "</li>\n",
       "<li><p>&quot;LR&quot; : Uses a logistic regression framework to determine differentially\n",
       "expressed genes. Constructs a logistic regression model predicting group\n",
       "membership based on each feature individually and compares this to a null\n",
       "model with a likelihood ratio test.\n",
       "</p>\n",
       "</li>\n",
       "<li><p>&quot;MAST&quot; : Identifies differentially expressed genes between two groups\n",
       "of cells using a hurdle model tailored to scRNA-seq data. Utilizes the MAST\n",
       "package to run the DE testing.\n",
       "</p>\n",
       "</li>\n",
       "<li><p>&quot;DESeq2&quot; : Identifies differentially expressed genes between two groups\n",
       "of cells based on a model using DESeq2 which uses a negative binomial\n",
       "distribution (Love et al, Genome Biology, 2014).This test does not support\n",
       "pre-filtering of genes based on average difference (or percent detection rate)\n",
       "between cell groups. However, genes may be pre-filtered based on their\n",
       "minimum detection rate (min.pct) across both cell groups. To use this method,\n",
       "please install DESeq2, using the instructions at\n",
       "https://bioconductor.org/packages/release/bioc/html/DESeq2.html\n",
       "</p>\n",
       "</li></ul>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>min.pct</code></td>\n",
       "<td>\n",
       "<p>only test genes that are detected in a minimum fraction of\n",
       "min.pct cells in either of the two populations. Meant to speed up the function\n",
       "by not testing genes that are very infrequently expressed. Default is 0.1</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>min.diff.pct</code></td>\n",
       "<td>\n",
       "<p>only test genes that show a minimum difference in the\n",
       "fraction of detection between the two groups. Set to -Inf by default</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>verbose</code></td>\n",
       "<td>\n",
       "<p>Print a progress bar once expression testing begins</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>only.pos</code></td>\n",
       "<td>\n",
       "<p>Only return positive markers (FALSE by default)</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>max.cells.per.ident</code></td>\n",
       "<td>\n",
       "<p>Down sample each identity class to a max number.\n",
       "Default is no downsampling. Not activated by default (set to Inf)</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>random.seed</code></td>\n",
       "<td>\n",
       "<p>Random seed for downsampling</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>latent.vars</code></td>\n",
       "<td>\n",
       "<p>Variables to test, used only when <code>test.use</code> is one of\n",
       "'LR', 'negbinom', 'poisson', or 'MAST'</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>min.cells.feature</code></td>\n",
       "<td>\n",
       "<p>Minimum number of cells expressing the feature in at least one\n",
       "of the two groups, currently only used for poisson and negative binomial tests</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>min.cells.group</code></td>\n",
       "<td>\n",
       "<p>Minimum number of cells in one of the groups</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>pseudocount.use</code></td>\n",
       "<td>\n",
       "<p>Pseudocount to add to averaged expression values when\n",
       "calculating logFC. 1 by default.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>fc.results</code></td>\n",
       "<td>\n",
       "<p>data.frame from FoldChange</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>densify</code></td>\n",
       "<td>\n",
       "<p>Convert the sparse matrix to a dense form before running the DE test. This can provide speedups but might require higher memory; default is FALSE</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>mean.fxn</code></td>\n",
       "<td>\n",
       "<p>Function to use for fold change or average difference calculation.\n",
       "If NULL, the appropriate function will be chose according to the slot used</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>fc.name</code></td>\n",
       "<td>\n",
       "<p>Name of the fold change, average difference, or custom function column\n",
       "in the output data.frame. If NULL, the fold change column will be named\n",
       "according to the logarithm base (eg, &quot;avg_log2FC&quot;), or if using the scale.data\n",
       "slot &quot;avg_diff&quot;.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>base</code></td>\n",
       "<td>\n",
       "<p>The base with respect to which logarithms are computed.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>ident.1</code></td>\n",
       "<td>\n",
       "<p>Identity class to define markers for; pass an object of class\n",
       "<code>phylo</code> or 'clustertree' to find markers for a node in a cluster tree;\n",
       "passing 'clustertree' requires <code>BuildClusterTree</code> to have been run</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>ident.2</code></td>\n",
       "<td>\n",
       "<p>A second identity class for comparison; if <code>NULL</code>,\n",
       "use all other cells for comparison; if an object of class <code>phylo</code> or\n",
       "'clustertree' is passed to <code>ident.1</code>, must pass a node to find markers for</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>group.by</code></td>\n",
       "<td>\n",
       "<p>Regroup cells into a different identity class prior to performing differential expression (see example)</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>subset.ident</code></td>\n",
       "<td>\n",
       "<p>Subset a particular identity class prior to regrouping. Only relevant if group.by is set (see example)</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>assay</code></td>\n",
       "<td>\n",
       "<p>Assay to use in differential expression testing</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>reduction</code></td>\n",
       "<td>\n",
       "<p>Reduction to use in differential expression testing - will test for DE on cell embeddings</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "\n",
       "\n",
       "<h3>Details</h3>\n",
       "\n",
       "<p>p-value adjustment is performed using bonferroni correction based on\n",
       "the total number of genes in the dataset. Other correction methods are not\n",
       "recommended, as Seurat pre-filters genes using the arguments above, reducing\n",
       "the number of tests performed. Lastly, as Aaron Lun has pointed out, p-values\n",
       "should be interpreted cautiously, as the genes used for clustering are the\n",
       "same genes tested for differential expression.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Value</h3>\n",
       "\n",
       "<p>data.frame with a ranked list of putative markers as rows, and associated\n",
       "statistics as columns (p-values, ROC score, etc., depending on the test used (<code>test.use</code>)). The following columns are always present:\n",
       "</p>\n",
       "\n",
       "<ul>\n",
       "<li> <p><code>avg_logFC</code>: log fold-chage of the average expression between the two groups. Positive values indicate that the gene is more highly expressed in the first group\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>pct.1</code>: The percentage of cells where the gene is detected in the first group\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>pct.2</code>: The percentage of cells where the gene is detected in the second group\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>p_val_adj</code>: Adjusted p-value, based on bonferroni correction using all genes in the dataset\n",
       "</p>\n",
       "</li></ul>\n",
       "\n",
       "\n",
       "\n",
       "<h3>References</h3>\n",
       "\n",
       "<p>McDavid A, Finak G, Chattopadyay PK, et al. Data exploration,\n",
       "quality control and testing in single-cell qPCR-based gene expression experiments.\n",
       "Bioinformatics. 2013;29(4):461-467. doi:10.1093/bioinformatics/bts714\n",
       "</p>\n",
       "<p>Trapnell C, et al. The dynamics and regulators of cell fate\n",
       "decisions are revealed by pseudotemporal ordering of single cells. Nature\n",
       "Biotechnology volume 32, pages 381-386 (2014)\n",
       "</p>\n",
       "<p>Andrew McDavid, Greg Finak and Masanao Yajima (2017). MAST: Model-based\n",
       "Analysis of Single Cell Transcriptomics. R package version 1.2.1.\n",
       "https://github.com/RGLab/MAST/\n",
       "</p>\n",
       "<p>Love MI, Huber W and Anders S (2014). &quot;Moderated estimation of\n",
       "fold change and dispersion for RNA-seq data with DESeq2.&quot; Genome Biology.\n",
       "https://bioconductor.org/packages/release/bioc/html/DESeq2.html\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>See Also</h3>\n",
       "\n",
       "<p><code>FoldChange</code>\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Examples</h3>\n",
       "\n",
       "<pre>\n",
       "data(\"pbmc_small\")\n",
       "# Find markers for cluster 2\n",
       "markers &lt;- FindMarkers(object = pbmc_small, ident.1 = 2)\n",
       "head(x = markers)\n",
       "\n",
       "# Take all cells in cluster 2, and find markers that separate cells in the 'g1' group (metadata\n",
       "# variable 'group')\n",
       "markers &lt;- FindMarkers(pbmc_small, ident.1 = \"g1\", group.by = 'groups', subset.ident = \"2\")\n",
       "head(x = markers)\n",
       "\n",
       "# Pass 'clustertree' or an object of class phylo to ident.1 and\n",
       "# a node to ident.2 as a replacement for FindMarkersNode\n",
       "pbmc_small &lt;- BuildClusterTree(object = pbmc_small)\n",
       "markers &lt;- FindMarkers(object = pbmc_small, ident.1 = 'clustertree', ident.2 = 5)\n",
       "head(x = markers)\n",
       "\n",
       "</pre>\n",
       "\n",
       "<hr /><div style=\"text-align: center;\">[Package <em>Seurat</em> version 4.0.4 ]</div>"
      ],
      "text/latex": [
       "\\inputencoding{utf8}\n",
       "\\HeaderA{FindMarkers}{Gene expression markers of identity classes}{FindMarkers}\n",
       "\\methaliasA{FindMarkers.Assay}{FindMarkers}{FindMarkers.Assay}\n",
       "\\methaliasA{FindMarkers.default}{FindMarkers}{FindMarkers.default}\n",
       "\\methaliasA{FindMarkers.DimReduc}{FindMarkers}{FindMarkers.DimReduc}\n",
       "\\methaliasA{FindMarkers.Seurat}{FindMarkers}{FindMarkers.Seurat}\n",
       "\\aliasA{FindMarkersNode}{FindMarkers}{FindMarkersNode}\n",
       "\\keyword{differential\\_expression}{FindMarkers}\n",
       "%\n",
       "\\begin{Description}\\relax\n",
       "Finds markers (differentially expressed genes) for identity classes\n",
       "\\end{Description}\n",
       "%\n",
       "\\begin{Usage}\n",
       "\\begin{verbatim}\n",
       "FindMarkers(object, ...)\n",
       "\n",
       "## Default S3 method:\n",
       "FindMarkers(\n",
       "  object,\n",
       "  slot = \"data\",\n",
       "  counts = numeric(),\n",
       "  cells.1 = NULL,\n",
       "  cells.2 = NULL,\n",
       "  features = NULL,\n",
       "  logfc.threshold = 0.25,\n",
       "  test.use = \"wilcox\",\n",
       "  min.pct = 0.1,\n",
       "  min.diff.pct = -Inf,\n",
       "  verbose = TRUE,\n",
       "  only.pos = FALSE,\n",
       "  max.cells.per.ident = Inf,\n",
       "  random.seed = 1,\n",
       "  latent.vars = NULL,\n",
       "  min.cells.feature = 3,\n",
       "  min.cells.group = 3,\n",
       "  pseudocount.use = 1,\n",
       "  fc.results = NULL,\n",
       "  densify = FALSE,\n",
       "  ...\n",
       ")\n",
       "\n",
       "## S3 method for class 'Assay'\n",
       "FindMarkers(\n",
       "  object,\n",
       "  slot = \"data\",\n",
       "  cells.1 = NULL,\n",
       "  cells.2 = NULL,\n",
       "  features = NULL,\n",
       "  logfc.threshold = 0.25,\n",
       "  test.use = \"wilcox\",\n",
       "  min.pct = 0.1,\n",
       "  min.diff.pct = -Inf,\n",
       "  verbose = TRUE,\n",
       "  only.pos = FALSE,\n",
       "  max.cells.per.ident = Inf,\n",
       "  random.seed = 1,\n",
       "  latent.vars = NULL,\n",
       "  min.cells.feature = 3,\n",
       "  min.cells.group = 3,\n",
       "  pseudocount.use = 1,\n",
       "  mean.fxn = NULL,\n",
       "  fc.name = NULL,\n",
       "  base = 2,\n",
       "  densify = FALSE,\n",
       "  ...\n",
       ")\n",
       "\n",
       "## S3 method for class 'DimReduc'\n",
       "FindMarkers(\n",
       "  object,\n",
       "  cells.1 = NULL,\n",
       "  cells.2 = NULL,\n",
       "  features = NULL,\n",
       "  logfc.threshold = 0.25,\n",
       "  test.use = \"wilcox\",\n",
       "  min.pct = 0.1,\n",
       "  min.diff.pct = -Inf,\n",
       "  verbose = TRUE,\n",
       "  only.pos = FALSE,\n",
       "  max.cells.per.ident = Inf,\n",
       "  random.seed = 1,\n",
       "  latent.vars = NULL,\n",
       "  min.cells.feature = 3,\n",
       "  min.cells.group = 3,\n",
       "  pseudocount.use = 1,\n",
       "  mean.fxn = rowMeans,\n",
       "  fc.name = NULL,\n",
       "  densify = FALSE,\n",
       "  ...\n",
       ")\n",
       "\n",
       "## S3 method for class 'Seurat'\n",
       "FindMarkers(\n",
       "  object,\n",
       "  ident.1 = NULL,\n",
       "  ident.2 = NULL,\n",
       "  group.by = NULL,\n",
       "  subset.ident = NULL,\n",
       "  assay = NULL,\n",
       "  slot = \"data\",\n",
       "  reduction = NULL,\n",
       "  features = NULL,\n",
       "  logfc.threshold = 0.25,\n",
       "  test.use = \"wilcox\",\n",
       "  min.pct = 0.1,\n",
       "  min.diff.pct = -Inf,\n",
       "  verbose = TRUE,\n",
       "  only.pos = FALSE,\n",
       "  max.cells.per.ident = Inf,\n",
       "  random.seed = 1,\n",
       "  latent.vars = NULL,\n",
       "  min.cells.feature = 3,\n",
       "  min.cells.group = 3,\n",
       "  pseudocount.use = 1,\n",
       "  mean.fxn = NULL,\n",
       "  fc.name = NULL,\n",
       "  base = 2,\n",
       "  densify = FALSE,\n",
       "  ...\n",
       ")\n",
       "\\end{verbatim}\n",
       "\\end{Usage}\n",
       "%\n",
       "\\begin{Arguments}\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{object}] An object\n",
       "\n",
       "\\item[\\code{...}] Arguments passed to other methods and to specific DE methods\n",
       "\n",
       "\\item[\\code{slot}] Slot to pull data from; note that if \\code{test.use} is \"negbinom\", \"poisson\", or \"DESeq2\",\n",
       "\\code{slot} will be set to \"counts\"\n",
       "\n",
       "\\item[\\code{counts}] Count matrix if using scale.data for DE tests. This is used for\n",
       "computing pct.1 and pct.2 and for filtering features based on fraction\n",
       "expressing\n",
       "\n",
       "\\item[\\code{cells.1}] Vector of cell names belonging to group 1\n",
       "\n",
       "\\item[\\code{cells.2}] Vector of cell names belonging to group 2\n",
       "\n",
       "\\item[\\code{features}] Genes to test. Default is to use all genes\n",
       "\n",
       "\\item[\\code{logfc.threshold}] Limit testing to genes which show, on average, at least\n",
       "X-fold difference (log-scale) between the two groups of cells. Default is 0.25\n",
       "Increasing logfc.threshold speeds up the function, but can miss weaker signals.\n",
       "\n",
       "\\item[\\code{test.use}] Denotes which test to use. Available options are:\n",
       "\\begin{itemize}\n",
       "\n",
       "\\item{} \"wilcox\" : Identifies differentially expressed genes between two\n",
       "groups of cells using a Wilcoxon Rank Sum test (default)\n",
       "\\item{} \"bimod\" : Likelihood-ratio test for single cell gene expression,\n",
       "(McDavid et al., Bioinformatics, 2013)\n",
       "\\item{} \"roc\" : Identifies 'markers' of gene expression using ROC analysis.\n",
       "For each gene, evaluates (using AUC) a classifier built on that gene alone,\n",
       "to classify between two groups of cells. An AUC value of 1 means that\n",
       "expression values for this gene alone can perfectly classify the two\n",
       "groupings (i.e. Each of the cells in cells.1 exhibit a higher level than\n",
       "each of the cells in cells.2). An AUC value of 0 also means there is perfect\n",
       "classification, but in the other direction. A value of 0.5 implies that\n",
       "the gene has no predictive power to classify the two groups. Returns a\n",
       "'predictive power' (abs(AUC-0.5) * 2) ranked matrix of putative differentially\n",
       "expressed genes.\n",
       "\\item{} \"t\" : Identify differentially expressed genes between two groups of\n",
       "cells using the Student's t-test.\n",
       "\\item{} \"negbinom\" : Identifies differentially expressed genes between two\n",
       "groups of cells using a negative binomial generalized linear model.\n",
       "Use only for UMI-based datasets\n",
       "\\item{} \"poisson\" : Identifies differentially expressed genes between two\n",
       "groups of cells using a poisson generalized linear model.\n",
       "Use only for UMI-based datasets\n",
       "\\item{} \"LR\" : Uses a logistic regression framework to determine differentially\n",
       "expressed genes. Constructs a logistic regression model predicting group\n",
       "membership based on each feature individually and compares this to a null\n",
       "model with a likelihood ratio test.\n",
       "\\item{} \"MAST\" : Identifies differentially expressed genes between two groups\n",
       "of cells using a hurdle model tailored to scRNA-seq data. Utilizes the MAST\n",
       "package to run the DE testing.\n",
       "\\item{} \"DESeq2\" : Identifies differentially expressed genes between two groups\n",
       "of cells based on a model using DESeq2 which uses a negative binomial\n",
       "distribution (Love et al, Genome Biology, 2014).This test does not support\n",
       "pre-filtering of genes based on average difference (or percent detection rate)\n",
       "between cell groups. However, genes may be pre-filtered based on their\n",
       "minimum detection rate (min.pct) across both cell groups. To use this method,\n",
       "please install DESeq2, using the instructions at\n",
       "https://bioconductor.org/packages/release/bioc/html/DESeq2.html\n",
       "\n",
       "\\end{itemize}\n",
       "\n",
       "\n",
       "\\item[\\code{min.pct}] only test genes that are detected in a minimum fraction of\n",
       "min.pct cells in either of the two populations. Meant to speed up the function\n",
       "by not testing genes that are very infrequently expressed. Default is 0.1\n",
       "\n",
       "\\item[\\code{min.diff.pct}] only test genes that show a minimum difference in the\n",
       "fraction of detection between the two groups. Set to -Inf by default\n",
       "\n",
       "\\item[\\code{verbose}] Print a progress bar once expression testing begins\n",
       "\n",
       "\\item[\\code{only.pos}] Only return positive markers (FALSE by default)\n",
       "\n",
       "\\item[\\code{max.cells.per.ident}] Down sample each identity class to a max number.\n",
       "Default is no downsampling. Not activated by default (set to Inf)\n",
       "\n",
       "\\item[\\code{random.seed}] Random seed for downsampling\n",
       "\n",
       "\\item[\\code{latent.vars}] Variables to test, used only when \\code{test.use} is one of\n",
       "'LR', 'negbinom', 'poisson', or 'MAST'\n",
       "\n",
       "\\item[\\code{min.cells.feature}] Minimum number of cells expressing the feature in at least one\n",
       "of the two groups, currently only used for poisson and negative binomial tests\n",
       "\n",
       "\\item[\\code{min.cells.group}] Minimum number of cells in one of the groups\n",
       "\n",
       "\\item[\\code{pseudocount.use}] Pseudocount to add to averaged expression values when\n",
       "calculating logFC. 1 by default.\n",
       "\n",
       "\\item[\\code{fc.results}] data.frame from FoldChange\n",
       "\n",
       "\\item[\\code{densify}] Convert the sparse matrix to a dense form before running the DE test. This can provide speedups but might require higher memory; default is FALSE\n",
       "\n",
       "\\item[\\code{mean.fxn}] Function to use for fold change or average difference calculation.\n",
       "If NULL, the appropriate function will be chose according to the slot used\n",
       "\n",
       "\\item[\\code{fc.name}] Name of the fold change, average difference, or custom function column\n",
       "in the output data.frame. If NULL, the fold change column will be named\n",
       "according to the logarithm base (eg, \"avg\\_log2FC\"), or if using the scale.data\n",
       "slot \"avg\\_diff\".\n",
       "\n",
       "\\item[\\code{base}] The base with respect to which logarithms are computed.\n",
       "\n",
       "\\item[\\code{ident.1}] Identity class to define markers for; pass an object of class\n",
       "\\code{phylo} or 'clustertree' to find markers for a node in a cluster tree;\n",
       "passing 'clustertree' requires \\code{\\LinkA{BuildClusterTree}{BuildClusterTree}} to have been run\n",
       "\n",
       "\\item[\\code{ident.2}] A second identity class for comparison; if \\code{NULL},\n",
       "use all other cells for comparison; if an object of class \\code{phylo} or\n",
       "'clustertree' is passed to \\code{ident.1}, must pass a node to find markers for\n",
       "\n",
       "\\item[\\code{group.by}] Regroup cells into a different identity class prior to performing differential expression (see example)\n",
       "\n",
       "\\item[\\code{subset.ident}] Subset a particular identity class prior to regrouping. Only relevant if group.by is set (see example)\n",
       "\n",
       "\\item[\\code{assay}] Assay to use in differential expression testing\n",
       "\n",
       "\\item[\\code{reduction}] Reduction to use in differential expression testing - will test for DE on cell embeddings\n",
       "\\end{ldescription}\n",
       "\\end{Arguments}\n",
       "%\n",
       "\\begin{Details}\\relax\n",
       "p-value adjustment is performed using bonferroni correction based on\n",
       "the total number of genes in the dataset. Other correction methods are not\n",
       "recommended, as Seurat pre-filters genes using the arguments above, reducing\n",
       "the number of tests performed. Lastly, as Aaron Lun has pointed out, p-values\n",
       "should be interpreted cautiously, as the genes used for clustering are the\n",
       "same genes tested for differential expression.\n",
       "\\end{Details}\n",
       "%\n",
       "\\begin{Value}\n",
       "data.frame with a ranked list of putative markers as rows, and associated\n",
       "statistics as columns (p-values, ROC score, etc., depending on the test used (\\code{test.use})). The following columns are always present:\n",
       "\\begin{itemize}\n",
       "\n",
       "\\item{} \\code{avg\\_logFC}: log fold-chage of the average expression between the two groups. Positive values indicate that the gene is more highly expressed in the first group\n",
       "\\item{} \\code{pct.1}: The percentage of cells where the gene is detected in the first group\n",
       "\\item{} \\code{pct.2}: The percentage of cells where the gene is detected in the second group\n",
       "\\item{} \\code{p\\_val\\_adj}: Adjusted p-value, based on bonferroni correction using all genes in the dataset\n",
       "\n",
       "\\end{itemize}\n",
       "\n",
       "\\end{Value}\n",
       "%\n",
       "\\begin{References}\\relax\n",
       "McDavid A, Finak G, Chattopadyay PK, et al. Data exploration,\n",
       "quality control and testing in single-cell qPCR-based gene expression experiments.\n",
       "Bioinformatics. 2013;29(4):461-467. doi:10.1093/bioinformatics/bts714\n",
       "\n",
       "Trapnell C, et al. The dynamics and regulators of cell fate\n",
       "decisions are revealed by pseudotemporal ordering of single cells. Nature\n",
       "Biotechnology volume 32, pages 381-386 (2014)\n",
       "\n",
       "Andrew McDavid, Greg Finak and Masanao Yajima (2017). MAST: Model-based\n",
       "Analysis of Single Cell Transcriptomics. R package version 1.2.1.\n",
       "https://github.com/RGLab/MAST/\n",
       "\n",
       "Love MI, Huber W and Anders S (2014). \"Moderated estimation of\n",
       "fold change and dispersion for RNA-seq data with DESeq2.\" Genome Biology.\n",
       "https://bioconductor.org/packages/release/bioc/html/DESeq2.html\n",
       "\\end{References}\n",
       "%\n",
       "\\begin{SeeAlso}\\relax\n",
       "\\code{FoldChange}\n",
       "\\end{SeeAlso}\n",
       "%\n",
       "\\begin{Examples}\n",
       "\\begin{ExampleCode}\n",
       "data(\"pbmc_small\")\n",
       "# Find markers for cluster 2\n",
       "markers <- FindMarkers(object = pbmc_small, ident.1 = 2)\n",
       "head(x = markers)\n",
       "\n",
       "# Take all cells in cluster 2, and find markers that separate cells in the 'g1' group (metadata\n",
       "# variable 'group')\n",
       "markers <- FindMarkers(pbmc_small, ident.1 = \"g1\", group.by = 'groups', subset.ident = \"2\")\n",
       "head(x = markers)\n",
       "\n",
       "# Pass 'clustertree' or an object of class phylo to ident.1 and\n",
       "# a node to ident.2 as a replacement for FindMarkersNode\n",
       "pbmc_small <- BuildClusterTree(object = pbmc_small)\n",
       "markers <- FindMarkers(object = pbmc_small, ident.1 = 'clustertree', ident.2 = 5)\n",
       "head(x = markers)\n",
       "\n",
       "\\end{ExampleCode}\n",
       "\\end{Examples}"
      ],
      "text/plain": [
       "FindMarkers               package:Seurat               R Documentation\n",
       "\n",
       "_\bG_\be_\bn_\be _\be_\bx_\bp_\br_\be_\bs_\bs_\bi_\bo_\bn _\bm_\ba_\br_\bk_\be_\br_\bs _\bo_\bf _\bi_\bd_\be_\bn_\bt_\bi_\bt_\by _\bc_\bl_\ba_\bs_\bs_\be_\bs\n",
       "\n",
       "_\bD_\be_\bs_\bc_\br_\bi_\bp_\bt_\bi_\bo_\bn:\n",
       "\n",
       "     Finds markers (differentially expressed genes) for identity\n",
       "     classes\n",
       "\n",
       "_\bU_\bs_\ba_\bg_\be:\n",
       "\n",
       "     FindMarkers(object, ...)\n",
       "     \n",
       "     ## Default S3 method:\n",
       "     FindMarkers(\n",
       "       object,\n",
       "       slot = \"data\",\n",
       "       counts = numeric(),\n",
       "       cells.1 = NULL,\n",
       "       cells.2 = NULL,\n",
       "       features = NULL,\n",
       "       logfc.threshold = 0.25,\n",
       "       test.use = \"wilcox\",\n",
       "       min.pct = 0.1,\n",
       "       min.diff.pct = -Inf,\n",
       "       verbose = TRUE,\n",
       "       only.pos = FALSE,\n",
       "       max.cells.per.ident = Inf,\n",
       "       random.seed = 1,\n",
       "       latent.vars = NULL,\n",
       "       min.cells.feature = 3,\n",
       "       min.cells.group = 3,\n",
       "       pseudocount.use = 1,\n",
       "       fc.results = NULL,\n",
       "       densify = FALSE,\n",
       "       ...\n",
       "     )\n",
       "     \n",
       "     ## S3 method for class 'Assay'\n",
       "     FindMarkers(\n",
       "       object,\n",
       "       slot = \"data\",\n",
       "       cells.1 = NULL,\n",
       "       cells.2 = NULL,\n",
       "       features = NULL,\n",
       "       logfc.threshold = 0.25,\n",
       "       test.use = \"wilcox\",\n",
       "       min.pct = 0.1,\n",
       "       min.diff.pct = -Inf,\n",
       "       verbose = TRUE,\n",
       "       only.pos = FALSE,\n",
       "       max.cells.per.ident = Inf,\n",
       "       random.seed = 1,\n",
       "       latent.vars = NULL,\n",
       "       min.cells.feature = 3,\n",
       "       min.cells.group = 3,\n",
       "       pseudocount.use = 1,\n",
       "       mean.fxn = NULL,\n",
       "       fc.name = NULL,\n",
       "       base = 2,\n",
       "       densify = FALSE,\n",
       "       ...\n",
       "     )\n",
       "     \n",
       "     ## S3 method for class 'DimReduc'\n",
       "     FindMarkers(\n",
       "       object,\n",
       "       cells.1 = NULL,\n",
       "       cells.2 = NULL,\n",
       "       features = NULL,\n",
       "       logfc.threshold = 0.25,\n",
       "       test.use = \"wilcox\",\n",
       "       min.pct = 0.1,\n",
       "       min.diff.pct = -Inf,\n",
       "       verbose = TRUE,\n",
       "       only.pos = FALSE,\n",
       "       max.cells.per.ident = Inf,\n",
       "       random.seed = 1,\n",
       "       latent.vars = NULL,\n",
       "       min.cells.feature = 3,\n",
       "       min.cells.group = 3,\n",
       "       pseudocount.use = 1,\n",
       "       mean.fxn = rowMeans,\n",
       "       fc.name = NULL,\n",
       "       densify = FALSE,\n",
       "       ...\n",
       "     )\n",
       "     \n",
       "     ## S3 method for class 'Seurat'\n",
       "     FindMarkers(\n",
       "       object,\n",
       "       ident.1 = NULL,\n",
       "       ident.2 = NULL,\n",
       "       group.by = NULL,\n",
       "       subset.ident = NULL,\n",
       "       assay = NULL,\n",
       "       slot = \"data\",\n",
       "       reduction = NULL,\n",
       "       features = NULL,\n",
       "       logfc.threshold = 0.25,\n",
       "       test.use = \"wilcox\",\n",
       "       min.pct = 0.1,\n",
       "       min.diff.pct = -Inf,\n",
       "       verbose = TRUE,\n",
       "       only.pos = FALSE,\n",
       "       max.cells.per.ident = Inf,\n",
       "       random.seed = 1,\n",
       "       latent.vars = NULL,\n",
       "       min.cells.feature = 3,\n",
       "       min.cells.group = 3,\n",
       "       pseudocount.use = 1,\n",
       "       mean.fxn = NULL,\n",
       "       fc.name = NULL,\n",
       "       base = 2,\n",
       "       densify = FALSE,\n",
       "       ...\n",
       "     )\n",
       "     \n",
       "_\bA_\br_\bg_\bu_\bm_\be_\bn_\bt_\bs:\n",
       "\n",
       "  object: An object\n",
       "\n",
       "     ...: Arguments passed to other methods and to specific DE methods\n",
       "\n",
       "    slot: Slot to pull data from; note that if ‘test.use’ is\n",
       "          \"negbinom\", \"poisson\", or \"DESeq2\", ‘slot’ will be set to\n",
       "          \"counts\"\n",
       "\n",
       "  counts: Count matrix if using scale.data for DE tests. This is used\n",
       "          for computing pct.1 and pct.2 and for filtering features\n",
       "          based on fraction expressing\n",
       "\n",
       " cells.1: Vector of cell names belonging to group 1\n",
       "\n",
       " cells.2: Vector of cell names belonging to group 2\n",
       "\n",
       "features: Genes to test. Default is to use all genes\n",
       "\n",
       "logfc.threshold: Limit testing to genes which show, on average, at\n",
       "          least X-fold difference (log-scale) between the two groups of\n",
       "          cells. Default is 0.25 Increasing logfc.threshold speeds up\n",
       "          the function, but can miss weaker signals.\n",
       "\n",
       "test.use: Denotes which test to use. Available options are:\n",
       "\n",
       "            • \"wilcox\" : Identifies differentially expressed genes\n",
       "              between two groups of cells using a Wilcoxon Rank Sum\n",
       "              test (default)\n",
       "\n",
       "            • \"bimod\" : Likelihood-ratio test for single cell gene\n",
       "              expression, (McDavid et al., Bioinformatics, 2013)\n",
       "\n",
       "            • \"roc\" : Identifies 'markers' of gene expression using ROC\n",
       "              analysis.  For each gene, evaluates (using AUC) a\n",
       "              classifier built on that gene alone, to classify between\n",
       "              two groups of cells. An AUC value of 1 means that\n",
       "              expression values for this gene alone can perfectly\n",
       "              classify the two groupings (i.e. Each of the cells in\n",
       "              cells.1 exhibit a higher level than each of the cells in\n",
       "              cells.2). An AUC value of 0 also means there is perfect\n",
       "              classification, but in the other direction. A value of\n",
       "              0.5 implies that the gene has no predictive power to\n",
       "              classify the two groups. Returns a 'predictive power'\n",
       "              (abs(AUC-0.5) * 2) ranked matrix of putative\n",
       "              differentially expressed genes.\n",
       "\n",
       "            • \"t\" : Identify differentially expressed genes between two\n",
       "              groups of cells using the Student's t-test.\n",
       "\n",
       "            • \"negbinom\" : Identifies differentially expressed genes\n",
       "              between two groups of cells using a negative binomial\n",
       "              generalized linear model.  Use only for UMI-based\n",
       "              datasets\n",
       "\n",
       "            • \"poisson\" : Identifies differentially expressed genes\n",
       "              between two groups of cells using a poisson generalized\n",
       "              linear model.  Use only for UMI-based datasets\n",
       "\n",
       "            • \"LR\" : Uses a logistic regression framework to determine\n",
       "              differentially expressed genes. Constructs a logistic\n",
       "              regression model predicting group membership based on\n",
       "              each feature individually and compares this to a null\n",
       "              model with a likelihood ratio test.\n",
       "\n",
       "            • \"MAST\" : Identifies differentially expressed genes\n",
       "              between two groups of cells using a hurdle model tailored\n",
       "              to scRNA-seq data. Utilizes the MAST package to run the\n",
       "              DE testing.\n",
       "\n",
       "            • \"DESeq2\" : Identifies differentially expressed genes\n",
       "              between two groups of cells based on a model using DESeq2\n",
       "              which uses a negative binomial distribution (Love et al,\n",
       "              Genome Biology, 2014).This test does not support\n",
       "              pre-filtering of genes based on average difference (or\n",
       "              percent detection rate) between cell groups. However,\n",
       "              genes may be pre-filtered based on their minimum\n",
       "              detection rate (min.pct) across both cell groups. To use\n",
       "              this method, please install DESeq2, using the\n",
       "              instructions at\n",
       "              https://bioconductor.org/packages/release/bioc/html/DESeq2.html\n",
       "\n",
       " min.pct: only test genes that are detected in a minimum fraction of\n",
       "          min.pct cells in either of the two populations. Meant to\n",
       "          speed up the function by not testing genes that are very\n",
       "          infrequently expressed. Default is 0.1\n",
       "\n",
       "min.diff.pct: only test genes that show a minimum difference in the\n",
       "          fraction of detection between the two groups. Set to -Inf by\n",
       "          default\n",
       "\n",
       " verbose: Print a progress bar once expression testing begins\n",
       "\n",
       "only.pos: Only return positive markers (FALSE by default)\n",
       "\n",
       "max.cells.per.ident: Down sample each identity class to a max number.\n",
       "          Default is no downsampling. Not activated by default (set to\n",
       "          Inf)\n",
       "\n",
       "random.seed: Random seed for downsampling\n",
       "\n",
       "latent.vars: Variables to test, used only when ‘test.use’ is one of\n",
       "          'LR', 'negbinom', 'poisson', or 'MAST'\n",
       "\n",
       "min.cells.feature: Minimum number of cells expressing the feature in at\n",
       "          least one of the two groups, currently only used for poisson\n",
       "          and negative binomial tests\n",
       "\n",
       "min.cells.group: Minimum number of cells in one of the groups\n",
       "\n",
       "pseudocount.use: Pseudocount to add to averaged expression values when\n",
       "          calculating logFC. 1 by default.\n",
       "\n",
       "fc.results: data.frame from FoldChange\n",
       "\n",
       " densify: Convert the sparse matrix to a dense form before running the\n",
       "          DE test. This can provide speedups but might require higher\n",
       "          memory; default is FALSE\n",
       "\n",
       "mean.fxn: Function to use for fold change or average difference\n",
       "          calculation. If NULL, the appropriate function will be chose\n",
       "          according to the slot used\n",
       "\n",
       " fc.name: Name of the fold change, average difference, or custom\n",
       "          function column in the output data.frame. If NULL, the fold\n",
       "          change column will be named according to the logarithm base\n",
       "          (eg, \"avg_log2FC\"), or if using the scale.data slot\n",
       "          \"avg_diff\".\n",
       "\n",
       "    base: The base with respect to which logarithms are computed.\n",
       "\n",
       " ident.1: Identity class to define markers for; pass an object of class\n",
       "          ‘phylo’ or 'clustertree' to find markers for a node in a\n",
       "          cluster tree; passing 'clustertree' requires\n",
       "          ‘BuildClusterTree’ to have been run\n",
       "\n",
       " ident.2: A second identity class for comparison; if ‘NULL’, use all\n",
       "          other cells for comparison; if an object of class ‘phylo’ or\n",
       "          'clustertree' is passed to ‘ident.1’, must pass a node to\n",
       "          find markers for\n",
       "\n",
       "group.by: Regroup cells into a different identity class prior to\n",
       "          performing differential expression (see example)\n",
       "\n",
       "subset.ident: Subset a particular identity class prior to regrouping.\n",
       "          Only relevant if group.by is set (see example)\n",
       "\n",
       "   assay: Assay to use in differential expression testing\n",
       "\n",
       "reduction: Reduction to use in differential expression testing - will\n",
       "          test for DE on cell embeddings\n",
       "\n",
       "_\bD_\be_\bt_\ba_\bi_\bl_\bs:\n",
       "\n",
       "     p-value adjustment is performed using bonferroni correction based\n",
       "     on the total number of genes in the dataset. Other correction\n",
       "     methods are not recommended, as Seurat pre-filters genes using the\n",
       "     arguments above, reducing the number of tests performed. Lastly,\n",
       "     as Aaron Lun has pointed out, p-values should be interpreted\n",
       "     cautiously, as the genes used for clustering are the same genes\n",
       "     tested for differential expression.\n",
       "\n",
       "_\bV_\ba_\bl_\bu_\be:\n",
       "\n",
       "     data.frame with a ranked list of putative markers as rows, and\n",
       "     associated statistics as columns (p-values, ROC score, etc.,\n",
       "     depending on the test used (‘test.use’)). The following columns\n",
       "     are always present:\n",
       "\n",
       "        • ‘avg_logFC’: log fold-chage of the average expression between\n",
       "          the two groups. Positive values indicate that the gene is\n",
       "          more highly expressed in the first group\n",
       "\n",
       "        • ‘pct.1’: The percentage of cells where the gene is detected\n",
       "          in the first group\n",
       "\n",
       "        • ‘pct.2’: The percentage of cells where the gene is detected\n",
       "          in the second group\n",
       "\n",
       "        • ‘p_val_adj’: Adjusted p-value, based on bonferroni correction\n",
       "          using all genes in the dataset\n",
       "\n",
       "_\bR_\be_\bf_\be_\br_\be_\bn_\bc_\be_\bs:\n",
       "\n",
       "     McDavid A, Finak G, Chattopadyay PK, et al. Data exploration,\n",
       "     quality control and testing in single-cell qPCR-based gene\n",
       "     expression experiments. Bioinformatics. 2013;29(4):461-467.\n",
       "     doi:10.1093/bioinformatics/bts714\n",
       "\n",
       "     Trapnell C, et al. The dynamics and regulators of cell fate\n",
       "     decisions are revealed by pseudotemporal ordering of single cells.\n",
       "     Nature Biotechnology volume 32, pages 381-386 (2014)\n",
       "\n",
       "     Andrew McDavid, Greg Finak and Masanao Yajima (2017). MAST:\n",
       "     Model-based Analysis of Single Cell Transcriptomics. R package\n",
       "     version 1.2.1. https://github.com/RGLab/MAST/\n",
       "\n",
       "     Love MI, Huber W and Anders S (2014). \"Moderated estimation of\n",
       "     fold change and dispersion for RNA-seq data with DESeq2.\" Genome\n",
       "     Biology.\n",
       "     https://bioconductor.org/packages/release/bioc/html/DESeq2.html\n",
       "\n",
       "_\bS_\be_\be _\bA_\bl_\bs_\bo:\n",
       "\n",
       "     ‘FoldChange’\n",
       "\n",
       "_\bE_\bx_\ba_\bm_\bp_\bl_\be_\bs:\n",
       "\n",
       "     data(\"pbmc_small\")\n",
       "     # Find markers for cluster 2\n",
       "     markers <- FindMarkers(object = pbmc_small, ident.1 = 2)\n",
       "     head(x = markers)\n",
       "     \n",
       "     # Take all cells in cluster 2, and find markers that separate cells in the 'g1' group (metadata\n",
       "     # variable 'group')\n",
       "     markers <- FindMarkers(pbmc_small, ident.1 = \"g1\", group.by = 'groups', subset.ident = \"2\")\n",
       "     head(x = markers)\n",
       "     \n",
       "     # Pass 'clustertree' or an object of class phylo to ident.1 and\n",
       "     # a node to ident.2 as a replacement for FindMarkersNode\n",
       "     pbmc_small <- BuildClusterTree(object = pbmc_small)\n",
       "     markers <- FindMarkers(object = pbmc_small, ident.1 = 'clustertree', ident.2 = 5)\n",
       "     head(x = markers)\n",
       "     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "?FindMarkers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "  ident.1 = NULL,\n",
    "  ident.2 = NULL,\n",
    "  group.by = NULL,\n",
    "  subset.ident = NULL,\n",
    "  assay = NULL,\n",
    "  slot = \"data\",\n",
    "  reduction = NULL,\n",
    "  features = NULL,\n",
    "  logfc.threshold = 0.25,\n",
    "  test.use = \"wilcox\",\n",
    "  min.pct = 0.1,\n",
    "  min.diff.pct = -Inf,\n",
    "  verbose = TRUE,\n",
    "  only.pos = FALSE,\n",
    "  max.cells.per.ident = Inf,\n",
    "  random.seed = 1,\n",
    "  latent.vars = NULL,\n",
    "  min.cells.feature = 3,\n",
    "  min.cells.group = 3,\n",
    "  pseudocount.use = 1,\n",
    "  mean.fxn = NULL,\n",
    "  fc.name = NULL,\n",
    "  base = 2,\n",
    "  densify = FALSE,"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-01-07T18:35:01.369099Z",
     "iopub.status.busy": "2022-01-07T18:35:01.368106Z",
     "iopub.status.idle": "2022-01-07T18:35:30.117917Z",
     "shell.execute_reply": "2022-01-07T18:35:30.116864Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Ery\"\n",
      "[1] \"CMP\"\n",
      "[1] \"ST-HSC.1\"\n",
      "[1] \"LT-HSC.1\"\n",
      "[1] \"GMP\"\n",
      "[1] \"MKP\"\n",
      "[1] \"ERP\"\n",
      "[1] \"LT-HSC.2\"\n",
      "[1] \"MEP\"\n",
      "[1] \"ST-HSC.2\"\n",
      "[1] \"MPP\"\n"
     ]
    }
   ],
   "source": [
    "merged_obj <- SetIdent(merged_obj, value = \"manual_label\")\n",
    "\n",
    "for (c in levels(merged_obj)){\n",
    "    tmp = subset(x = merged_obj, subset = manual_label == c)\n",
    "    degs = FindMarkers(tmp, verbose=T,slot=\"data\",ident.1 = \"mNFIXko\", ident.2 = \"mWT\",\n",
    "            group.by=\"orig.ident\",min.pct=0,logfc.threshold = 0,test.used=\"wilcox\",pseudocount.use=1,features =deg)\n",
    "\n",
    "    print (c)\n",
    "    write.table(degs,paste(\"Cluster\",c,\".DEG.all.for_heatmap.tsv\",sep=\"\"),sep=\"\\t\")\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-01-07T18:37:39.381247Z",
     "iopub.status.busy": "2022-01-07T18:37:39.380259Z",
     "iopub.status.idle": "2022-01-07T18:37:39.390391Z",
     "shell.execute_reply": "2022-01-07T18:37:39.389421Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>530</li><li>5</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 530\n",
       "\\item 5\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 530\n",
       "2. 5\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 530   5"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dim(degs)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
