{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7d5db65f-ca4b-45f8-9b29-7cd6e3ff99a9",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-01T19:39:24.057493Z",
     "iopub.status.busy": "2021-09-01T19:39:24.057309Z",
     "iopub.status.idle": "2021-09-01T19:39:28.532608Z",
     "shell.execute_reply": "2021-09-01T19:39:28.531945Z",
     "shell.execute_reply.started": "2021-09-01T19:39:24.057455Z"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d4201557-e236-48cc-acfa-0ca47b06d986",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-01T19:39:32.135297Z",
     "iopub.status.busy": "2021-09-01T19:39:32.135023Z",
     "iopub.status.idle": "2021-09-01T19:39:32.313964Z",
     "shell.execute_reply": "2021-09-01T19:39:32.313338Z",
     "shell.execute_reply.started": "2021-09-01T19:39:32.135277Z"
    }
   },
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"https://raw.githubusercontent.com/scRNA-tools/scRNA-tools/master/database/tools.tsv\",sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8a7b30bd-a116-46e0-9d1d-8a13371da351",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-01T19:39:34.379311Z",
     "iopub.status.busy": "2021-09-01T19:39:34.378910Z",
     "iopub.status.idle": "2021-09-01T19:39:34.416326Z",
     "shell.execute_reply": "2021-09-01T19:39:34.415788Z",
     "shell.execute_reply.started": "2021-09-01T19:39:34.379294Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Tool</th>\n",
       "      <th>Platform</th>\n",
       "      <th>Code</th>\n",
       "      <th>Description</th>\n",
       "      <th>License</th>\n",
       "      <th>Added</th>\n",
       "      <th>Updated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>acorde</td>\n",
       "      <td>R</td>\n",
       "      <td>https://github.com/ConesaLab/acorde</td>\n",
       "      <td>Isoform co-usage networks from single-cell RNA...</td>\n",
       "      <td>GPL-3.0</td>\n",
       "      <td>2021-05-14</td>\n",
       "      <td>2021-05-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACTINN</td>\n",
       "      <td>Python</td>\n",
       "      <td>https://github.com/mafeiyang/ACTINN</td>\n",
       "      <td>ACTINN (Automated Cell Type Identification usi...</td>\n",
       "      <td>GPL-3.0</td>\n",
       "      <td>2019-02-01</td>\n",
       "      <td>2021-06-28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACTION</td>\n",
       "      <td>C++/R/MATLAB</td>\n",
       "      <td>http://compbio.mit.edu/ACTION</td>\n",
       "      <td>ACTION infers the functional identity of cells...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2018-04-23</td>\n",
       "      <td>2021-06-28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACTIONet</td>\n",
       "      <td>R/C++</td>\n",
       "      <td>https://github.com/shmohammadi86/ACTIONet</td>\n",
       "      <td>Robust multi-resolution analysis of single-cel...</td>\n",
       "      <td>GPL-2.0-or-later</td>\n",
       "      <td>2019-12-10</td>\n",
       "      <td>2021-06-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACTIVA</td>\n",
       "      <td>Python</td>\n",
       "      <td>https://github.com/SindiLab/ACTIVA</td>\n",
       "      <td>Realistic scRNAseq Generation with Automatic C...</td>\n",
       "      <td>MIT</td>\n",
       "      <td>2021-02-05</td>\n",
       "      <td>2021-02-05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Tool      Platform                                       Code  \\\n",
       "0    acorde             R        https://github.com/ConesaLab/acorde   \n",
       "1    ACTINN        Python        https://github.com/mafeiyang/ACTINN   \n",
       "2    ACTION  C++/R/MATLAB              http://compbio.mit.edu/ACTION   \n",
       "3  ACTIONet         R/C++  https://github.com/shmohammadi86/ACTIONet   \n",
       "4    ACTIVA        Python         https://github.com/SindiLab/ACTIVA   \n",
       "\n",
       "                                         Description           License  \\\n",
       "0  Isoform co-usage networks from single-cell RNA...           GPL-3.0   \n",
       "1  ACTINN (Automated Cell Type Identification usi...           GPL-3.0   \n",
       "2  ACTION infers the functional identity of cells...               NaN   \n",
       "3  Robust multi-resolution analysis of single-cel...  GPL-2.0-or-later   \n",
       "4  Realistic scRNAseq Generation with Automatic C...               MIT   \n",
       "\n",
       "        Added     Updated  \n",
       "0  2021-05-14  2021-05-14  \n",
       "1  2019-02-01  2021-06-28  \n",
       "2  2018-04-23  2021-06-28  \n",
       "3  2019-12-10  2021-06-11  \n",
       "4  2021-02-05  2021-02-05  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d07da084-c5a8-4bba-a280-f943558a43fa",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-01T19:39:47.838963Z",
     "iopub.status.busy": "2021-09-01T19:39:47.838758Z",
     "iopub.status.idle": "2021-09-01T19:39:47.843055Z",
     "shell.execute_reply": "2021-09-01T19:39:47.842335Z",
     "shell.execute_reply.started": "2021-09-01T19:39:47.838946Z"
    }
   },
   "outputs": [],
   "source": [
    "df = df.sort_values(\"Added\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "34c93f90-31e3-4269-87d6-1a97de6b64f8",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-01T19:39:49.267351Z",
     "iopub.status.busy": "2021-09-01T19:39:49.267191Z",
     "iopub.status.idle": "2021-09-01T19:39:49.276596Z",
     "shell.execute_reply": "2021-09-01T19:39:49.276115Z",
     "shell.execute_reply.started": "2021-09-01T19:39:49.267336Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Tool</th>\n",
       "      <th>Platform</th>\n",
       "      <th>Code</th>\n",
       "      <th>Description</th>\n",
       "      <th>License</th>\n",
       "      <th>Added</th>\n",
       "      <th>Updated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>BPSC</td>\n",
       "      <td>R</td>\n",
       "      <td>https://github.com/nghiavtr/BPSC</td>\n",
       "      <td>Beta-Poisson model for single-cell RNA-seq dat...</td>\n",
       "      <td>GPL-3.0</td>\n",
       "      <td>2016-09-08</td>\n",
       "      <td>2016-09-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>739</th>\n",
       "      <td>scLVM</td>\n",
       "      <td>R/Python</td>\n",
       "      <td>https://github.com/PMBio/scLVM</td>\n",
       "      <td>scLVM is a modelling framework for single-cell...</td>\n",
       "      <td>Apache-2.0</td>\n",
       "      <td>2016-09-08</td>\n",
       "      <td>2016-12-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>684</th>\n",
       "      <td>SCell</td>\n",
       "      <td>MATLAB</td>\n",
       "      <td>https://github.com/diazlab/SCell</td>\n",
       "      <td>SCell is an integrated software tool for quali...</td>\n",
       "      <td>GPL-1.0</td>\n",
       "      <td>2016-09-08</td>\n",
       "      <td>2016-09-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>901</th>\n",
       "      <td>SinQC</td>\n",
       "      <td>R/Python</td>\n",
       "      <td>http://www.morgridge.net/SinQC.html</td>\n",
       "      <td>A Method and Tool to Control Single-cell RNA-s...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2016-09-08</td>\n",
       "      <td>2017-04-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>507</th>\n",
       "      <td>Ouija</td>\n",
       "      <td>R</td>\n",
       "      <td>https://github.com/kieranrcampbell/ouija</td>\n",
       "      <td>Incorporate prior information into single-cell...</td>\n",
       "      <td>GPL-3.0-or-later</td>\n",
       "      <td>2016-09-08</td>\n",
       "      <td>2018-06-27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>677</th>\n",
       "      <td>scDiffCom</td>\n",
       "      <td>R</td>\n",
       "      <td>https://github.com/CyrilLagger/scDiffCom</td>\n",
       "      <td>scDiffCom infers cell type to cell type commun...</td>\n",
       "      <td>MIT</td>\n",
       "      <td>2021-08-27</td>\n",
       "      <td>2021-08-27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>715</th>\n",
       "      <td>scHPL</td>\n",
       "      <td>Python</td>\n",
       "      <td>https://github.com/lcmmichielsen/scHPL</td>\n",
       "      <td>A hierarchical progressive learning method whi...</td>\n",
       "      <td>MIT</td>\n",
       "      <td>2021-08-27</td>\n",
       "      <td>2021-08-27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>818</th>\n",
       "      <td>SCSit</td>\n",
       "      <td>C/Python</td>\n",
       "      <td>https://github.com/shang-qian/SCSit</td>\n",
       "      <td>A high-efficiency preprocessing tool for singl...</td>\n",
       "      <td>MIT</td>\n",
       "      <td>2021-08-27</td>\n",
       "      <td>2021-08-27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330</th>\n",
       "      <td>graph-sc</td>\n",
       "      <td>Python</td>\n",
       "      <td>https://github.com/ciortanmadalina/graph-sc-pa...</td>\n",
       "      <td>A method modeling scRNA-seq data as a graph, p...</td>\n",
       "      <td>MIT</td>\n",
       "      <td>2021-08-27</td>\n",
       "      <td>2021-08-27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>734</th>\n",
       "      <td>Sciviewer</td>\n",
       "      <td>Python</td>\n",
       "      <td>https://github.com/colabobio/sciviewer</td>\n",
       "      <td>This is an interactive viewer for 2D embedding...</td>\n",
       "      <td>MIT</td>\n",
       "      <td>2021-08-27</td>\n",
       "      <td>2021-08-27</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1033 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Tool  Platform                                               Code  \\\n",
       "58        BPSC         R                   https://github.com/nghiavtr/BPSC   \n",
       "739      scLVM  R/Python                     https://github.com/PMBio/scLVM   \n",
       "684      SCell    MATLAB                   https://github.com/diazlab/SCell   \n",
       "901      SinQC  R/Python                http://www.morgridge.net/SinQC.html   \n",
       "507      Ouija         R           https://github.com/kieranrcampbell/ouija   \n",
       "..         ...       ...                                                ...   \n",
       "677  scDiffCom         R           https://github.com/CyrilLagger/scDiffCom   \n",
       "715      scHPL    Python             https://github.com/lcmmichielsen/scHPL   \n",
       "818      SCSit  C/Python                https://github.com/shang-qian/SCSit   \n",
       "330   graph-sc    Python  https://github.com/ciortanmadalina/graph-sc-pa...   \n",
       "734  Sciviewer    Python             https://github.com/colabobio/sciviewer   \n",
       "\n",
       "                                           Description           License  \\\n",
       "58   Beta-Poisson model for single-cell RNA-seq dat...           GPL-3.0   \n",
       "739  scLVM is a modelling framework for single-cell...        Apache-2.0   \n",
       "684  SCell is an integrated software tool for quali...           GPL-1.0   \n",
       "901  A Method and Tool to Control Single-cell RNA-s...               NaN   \n",
       "507  Incorporate prior information into single-cell...  GPL-3.0-or-later   \n",
       "..                                                 ...               ...   \n",
       "677  scDiffCom infers cell type to cell type commun...               MIT   \n",
       "715  A hierarchical progressive learning method whi...               MIT   \n",
       "818  A high-efficiency preprocessing tool for singl...               MIT   \n",
       "330  A method modeling scRNA-seq data as a graph, p...               MIT   \n",
       "734  This is an interactive viewer for 2D embedding...               MIT   \n",
       "\n",
       "          Added     Updated  \n",
       "58   2016-09-08  2016-09-08  \n",
       "739  2016-09-08  2016-12-08  \n",
       "684  2016-09-08  2016-09-08  \n",
       "901  2016-09-08  2017-04-18  \n",
       "507  2016-09-08  2018-06-27  \n",
       "..          ...         ...  \n",
       "677  2021-08-27  2021-08-27  \n",
       "715  2021-08-27  2021-08-27  \n",
       "818  2021-08-27  2021-08-27  \n",
       "330  2021-08-27  2021-08-27  \n",
       "734  2021-08-27  2021-08-27  \n",
       "\n",
       "[1033 rows x 7 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a65a0e80-0be4-4076-949d-f4058d8e7d06",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-01T19:40:21.954764Z",
     "iopub.status.busy": "2021-09-01T19:40:21.954563Z",
     "iopub.status.idle": "2021-09-01T19:40:21.969203Z",
     "shell.execute_reply": "2021-09-01T19:40:21.968687Z",
     "shell.execute_reply.started": "2021-09-01T19:40:21.954748Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "df[df.Platform==\"R\"].to_csv(\"tools.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbb92523-0ffe-4213-ba54-79eae37ef56b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
