{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "df8ebc0d-6a65-4f1c-8110-7b175f888421",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-17T14:36:04.760847Z",
     "iopub.status.busy": "2021-09-17T14:36:04.758647Z",
     "iopub.status.idle": "2021-09-17T14:36:38.453014Z",
     "shell.execute_reply": "2021-09-17T14:36:38.451893Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Attaching SeuratObject\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘glue’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘cowplot’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:patchwork’:\n",
      "\n",
      "    align_plots\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"barcodes.tsv.gz\" \"features.tsv.gz\" \"matrix.mtx.gz\"  \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "10X data contains more than one type and is being returned as a list containing matrices of each type.\n",
      "\n",
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"barcodes.tsv.gz\" \"features.tsv.gz\" \"matrix.mtx.gz\"  \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "10X data contains more than one type and is being returned as a list containing matrices of each type.\n",
      "\n",
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n"
     ]
    }
   ],
   "source": [
    "library(Seurat)\n",
    "\n",
    "library(dplyr)\n",
    "library(glue)\n",
    "library(ggplot2)\n",
    "library(patchwork)\n",
    "library(cowplot)\n",
    "library(stringr)\n",
    "library(tidyr)\n",
    "library(\"dplyr\")\n",
    "\n",
    "sample_id = c(\"mWT\",\"mNFIXko\")\n",
    "myList = list()\n",
    "for(i in 1:length(sample_id)){\n",
    "    s=sample_id[i]\n",
    "    dir = glue(\"{s}/outs/filtered_feature_bc_matrix\")\n",
    "    print (list.files(dir))\n",
    "    myData <- Read10X(data.dir = dir)\n",
    "    myList[[i]]  <- CreateSeuratObject(counts = myData[[\"Gene Expression\"]], min.cells = 0, min.features = 0,project = s, assay = \"RNA\")\n",
    "    myList[[i]][[\"ADT\"]] <- CreateAssayObject(myData[[\"Antibody Capture\"]][, colnames(x = myList[[i]] )])\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1b3c6b8e-6cfc-4701-9dde-1e9491dac0f0",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-17T14:36:54.798469Z",
     "iopub.status.busy": "2021-09-17T14:36:54.778144Z",
     "iopub.status.idle": "2021-09-17T14:36:55.228015Z",
     "shell.execute_reply": "2021-09-17T14:36:55.226573Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table width=\"100%\" summary=\"page for Read10X {Seurat}\"><tr><td>Read10X {Seurat}</td><td style=\"text-align: right;\">R Documentation</td></tr></table>\n",
       "\n",
       "<h2>Load in data from 10X</h2>\n",
       "\n",
       "<h3>Description</h3>\n",
       "\n",
       "<p>Enables easy loading of sparse data matrices provided by 10X genomics.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Usage</h3>\n",
       "\n",
       "<pre>\n",
       "Read10X(\n",
       "  data.dir,\n",
       "  gene.column = 2,\n",
       "  cell.column = 1,\n",
       "  unique.features = TRUE,\n",
       "  strip.suffix = FALSE\n",
       ")\n",
       "</pre>\n",
       "\n",
       "\n",
       "<h3>Arguments</h3>\n",
       "\n",
       "<table summary=\"R argblock\">\n",
       "<tr valign=\"top\"><td><code>data.dir</code></td>\n",
       "<td>\n",
       "<p>Directory containing the matrix.mtx, genes.tsv (or features.tsv), and barcodes.tsv\n",
       "files provided by 10X. A vector or named vector can be given in order to load\n",
       "several data directories. If a named vector is given, the cell barcode names\n",
       "will be prefixed with the name.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>gene.column</code></td>\n",
       "<td>\n",
       "<p>Specify which column of genes.tsv or features.tsv to use for gene names; default is 2</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>cell.column</code></td>\n",
       "<td>\n",
       "<p>Specify which column of barcodes.tsv to use for cell names; default is 1</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>unique.features</code></td>\n",
       "<td>\n",
       "<p>Make feature names unique (default TRUE)</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>strip.suffix</code></td>\n",
       "<td>\n",
       "<p>Remove trailing &quot;-1&quot; if present in all cell barcodes.</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "\n",
       "\n",
       "<h3>Value</h3>\n",
       "\n",
       "<p>If features.csv indicates the data has multiple data types, a list\n",
       "containing a sparse matrix of the data from each type will be returned.\n",
       "Otherwise a sparse matrix containing the expression data will be returned.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Examples</h3>\n",
       "\n",
       "<pre>\n",
       "## Not run: \n",
       "# For output from CellRanger &lt; 3.0\n",
       "data_dir &lt;- 'path/to/data/directory'\n",
       "list.files(data_dir) # Should show barcodes.tsv, genes.tsv, and matrix.mtx\n",
       "expression_matrix &lt;- Read10X(data.dir = data_dir)\n",
       "seurat_object = CreateSeuratObject(counts = expression_matrix)\n",
       "\n",
       "# For output from CellRanger &gt;= 3.0 with multiple data types\n",
       "data_dir &lt;- 'path/to/data/directory'\n",
       "list.files(data_dir) # Should show barcodes.tsv.gz, features.tsv.gz, and matrix.mtx.gz\n",
       "data &lt;- Read10X(data.dir = data_dir)\n",
       "seurat_object = CreateSeuratObject(counts = data$`Gene Expression`)\n",
       "seurat_object[['Protein']] = CreateAssayObject(counts = data$`Antibody Capture`)\n",
       "\n",
       "## End(Not run)\n",
       "\n",
       "</pre>\n",
       "\n",
       "<hr /><div style=\"text-align: center;\">[Package <em>Seurat</em> version 4.0.4 ]</div>"
      ],
      "text/latex": [
       "\\inputencoding{utf8}\n",
       "\\HeaderA{Read10X}{Load in data from 10X}{Read10X}\n",
       "\\keyword{preprocessing}{Read10X}\n",
       "%\n",
       "\\begin{Description}\\relax\n",
       "Enables easy loading of sparse data matrices provided by 10X genomics.\n",
       "\\end{Description}\n",
       "%\n",
       "\\begin{Usage}\n",
       "\\begin{verbatim}\n",
       "Read10X(\n",
       "  data.dir,\n",
       "  gene.column = 2,\n",
       "  cell.column = 1,\n",
       "  unique.features = TRUE,\n",
       "  strip.suffix = FALSE\n",
       ")\n",
       "\\end{verbatim}\n",
       "\\end{Usage}\n",
       "%\n",
       "\\begin{Arguments}\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{data.dir}] Directory containing the matrix.mtx, genes.tsv (or features.tsv), and barcodes.tsv\n",
       "files provided by 10X. A vector or named vector can be given in order to load\n",
       "several data directories. If a named vector is given, the cell barcode names\n",
       "will be prefixed with the name.\n",
       "\n",
       "\\item[\\code{gene.column}] Specify which column of genes.tsv or features.tsv to use for gene names; default is 2\n",
       "\n",
       "\\item[\\code{cell.column}] Specify which column of barcodes.tsv to use for cell names; default is 1\n",
       "\n",
       "\\item[\\code{unique.features}] Make feature names unique (default TRUE)\n",
       "\n",
       "\\item[\\code{strip.suffix}] Remove trailing \"-1\" if present in all cell barcodes.\n",
       "\\end{ldescription}\n",
       "\\end{Arguments}\n",
       "%\n",
       "\\begin{Value}\n",
       "If features.csv indicates the data has multiple data types, a list\n",
       "containing a sparse matrix of the data from each type will be returned.\n",
       "Otherwise a sparse matrix containing the expression data will be returned.\n",
       "\\end{Value}\n",
       "%\n",
       "\\begin{Examples}\n",
       "\\begin{ExampleCode}\n",
       "## Not run: \n",
       "# For output from CellRanger < 3.0\n",
       "data_dir <- 'path/to/data/directory'\n",
       "list.files(data_dir) # Should show barcodes.tsv, genes.tsv, and matrix.mtx\n",
       "expression_matrix <- Read10X(data.dir = data_dir)\n",
       "seurat_object = CreateSeuratObject(counts = expression_matrix)\n",
       "\n",
       "# For output from CellRanger >= 3.0 with multiple data types\n",
       "data_dir <- 'path/to/data/directory'\n",
       "list.files(data_dir) # Should show barcodes.tsv.gz, features.tsv.gz, and matrix.mtx.gz\n",
       "data <- Read10X(data.dir = data_dir)\n",
       "seurat_object = CreateSeuratObject(counts = data$`Gene Expression`)\n",
       "seurat_object[['Protein']] = CreateAssayObject(counts = data$`Antibody Capture`)\n",
       "\n",
       "## End(Not run)\n",
       "\n",
       "\\end{ExampleCode}\n",
       "\\end{Examples}"
      ],
      "text/plain": [
       "Read10X                 package:Seurat                 R Documentation\n",
       "\n",
       "_\bL_\bo_\ba_\bd _\bi_\bn _\bd_\ba_\bt_\ba _\bf_\br_\bo_\bm _\b1_\b0_\bX\n",
       "\n",
       "_\bD_\be_\bs_\bc_\br_\bi_\bp_\bt_\bi_\bo_\bn:\n",
       "\n",
       "     Enables easy loading of sparse data matrices provided by 10X\n",
       "     genomics.\n",
       "\n",
       "_\bU_\bs_\ba_\bg_\be:\n",
       "\n",
       "     Read10X(\n",
       "       data.dir,\n",
       "       gene.column = 2,\n",
       "       cell.column = 1,\n",
       "       unique.features = TRUE,\n",
       "       strip.suffix = FALSE\n",
       "     )\n",
       "     \n",
       "_\bA_\br_\bg_\bu_\bm_\be_\bn_\bt_\bs:\n",
       "\n",
       "data.dir: Directory containing the matrix.mtx, genes.tsv (or\n",
       "          features.tsv), and barcodes.tsv files provided by 10X. A\n",
       "          vector or named vector can be given in order to load several\n",
       "          data directories. If a named vector is given, the cell\n",
       "          barcode names will be prefixed with the name.\n",
       "\n",
       "gene.column: Specify which column of genes.tsv or features.tsv to use\n",
       "          for gene names; default is 2\n",
       "\n",
       "cell.column: Specify which column of barcodes.tsv to use for cell\n",
       "          names; default is 1\n",
       "\n",
       "unique.features: Make feature names unique (default TRUE)\n",
       "\n",
       "strip.suffix: Remove trailing \"-1\" if present in all cell barcodes.\n",
       "\n",
       "_\bV_\ba_\bl_\bu_\be:\n",
       "\n",
       "     If features.csv indicates the data has multiple data types, a list\n",
       "     containing a sparse matrix of the data from each type will be\n",
       "     returned.  Otherwise a sparse matrix containing the expression\n",
       "     data will be returned.\n",
       "\n",
       "_\bE_\bx_\ba_\bm_\bp_\bl_\be_\bs:\n",
       "\n",
       "     ## Not run:\n",
       "     \n",
       "     # For output from CellRanger < 3.0\n",
       "     data_dir <- 'path/to/data/directory'\n",
       "     list.files(data_dir) # Should show barcodes.tsv, genes.tsv, and matrix.mtx\n",
       "     expression_matrix <- Read10X(data.dir = data_dir)\n",
       "     seurat_object = CreateSeuratObject(counts = expression_matrix)\n",
       "     \n",
       "     # For output from CellRanger >= 3.0 with multiple data types\n",
       "     data_dir <- 'path/to/data/directory'\n",
       "     list.files(data_dir) # Should show barcodes.tsv.gz, features.tsv.gz, and matrix.mtx.gz\n",
       "     data <- Read10X(data.dir = data_dir)\n",
       "     seurat_object = CreateSeuratObject(counts = data$`Gene Expression`)\n",
       "     seurat_object[['Protein']] = CreateAssayObject(counts = data$`Antibody Capture`)\n",
       "     ## End(Not run)\n",
       "     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "?Read10X"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
